src/main/java/com/iota/iri/hash/Kerl.java

```diff
@@ -6,142 +6,174 @@
+                    power = 0;
+                    break;
+                case -1:
+                    value = value.multiply(RADIX_POWERS[power + 1]).subtract(BigInteger.ONE);
+                    power = 0;
+                    break;
+                default:
+                    throw new RuntimeException("unreachable");
+            }
+        }
+        // do leftover pow
+        if (power > 0) {
+            value = value.multiply(RADIX_POWERS[power]);
         }
-
         return value;
     }
 
     public static BigInteger bigIntFromBytes(final byte[] bytes, final int offset, final int size) {
-
-        return new BigInteger(Arrays.copyOfRange(bytes, offset, offset + size));
+        return (bytes.length == offset + size) ? new BigInteger(bytes)
+            : new BigInteger(Arrays.copyOfRange(bytes, offset, offset + size));
     }
 
-    public static void tritsFromBigInt(final BigInteger value, int[] destination, int offset, int size) {
+    public static void tritsFromBigInt(final BigInteger value, final int[] destination, final int offset, final int size) {
 
-        if(destination.length - offset < size) {
+        if (destination.length - offset < size) {
             throw new IllegalArgumentException("Destination array has invalid size");
         }
-
-        BigInteger absoluteValue = value.compareTo(BigInteger.ZERO) < 0 ? value.negate() : value;
+        final int signum = value.signum();
+        if (signum == 0) {
+            Arrays.fill(destination, offset, size, 0);
+            return;
+        }
+        BigInteger absoluteValue = value.abs();
         for (int i = 0; i < size; i++) {
-
-            BigInteger[] divRemainder = absoluteValue.divideAndRemainder(BigInteger.valueOf(Converter.RADIX));
-            int remainder = divRemainder[1].intValue();
+            BigInteger[] divRemainder = absoluteValue.divideAndRemainder(RADIX);
             absoluteValue = divRemainder[0];
 
+            int remainder = divRemainder[1].intValue();
             if (remainder > Converter.MAX_TRIT_VALUE) {
-
                 remainder = Converter.MIN_TRIT_VALUE;
                 absoluteValue = absoluteValue.add(BigInteger.ONE);
             }
-            destination[offset  + i] = remainder;
-        }
-
-        if (value.compareTo(BigInteger.ZERO) < 0) {
-
-            for (int i = 0; i < size; i++) {
-
-                destination[offset + i] = -destination[offset + i];
-            }
+            destination[offset + i] = signum < 0 ? -remainder : remainder;
         }
     }
-    public static void bytesFromBigInt(final BigInteger value, byte[] destination, int offset) {
-        if(destination.length - offset < BYTE_HASH_LENGTH) {
-            throw new IllegalArgumentException("Destination array has invalid size.");
-        }
-
-        final byte[] bytes = value.toByteArray();
-        int i = 0;
-        while (i + bytes.length < BYTE_HASH_LENGTH) {
 
```

Comment:
offset was never used - so I removed.
Also, Think it is clearer with Arrays.fill .. and also use some math to calc length without creating a variable for array.